udata/harvest/backends/dcat.py

```diff
@@ -188,37 +218,110 @@ def parse_graph(self, url: str, fmt: str) -> List[Graph]:
                     kwargs['type'] = 'uriref' if isinstance(node, URIRef) else 'blank'
                     self.add_item(id, **kwargs)
             graphs.append(graph)
+
+            next_record = self.next_record_if_should_continue(start, search_results)
+            if not next_record:
+                break
+
+            start = next_record
             page += 1
 
-            next_record = int(search_results.attrib['nextRecord'])
-            matched_count = int(search_results.attrib['numberOfRecordsMatched'])
-            returned_count = int(search_results.attrib['numberOfRecordsReturned'])
+            tree = lET.fromstring(
+                self.post(url, data=body.format(start=start, schema=self.DCAT_SCHEMA),
+                          headers=headers).content)
 
-            # Break conditions copied gratefully from
-            # noqa https://github.com/geonetwork/core-geonetwork/blob/main/harvesters/src/main/java/org/fao/geonet/kernel/harvest/harvester/csw/Harvester.java#L338-L369
-            break_conditions = (
-                # standard CSW: A value of 0 means all records have been returned.
-                next_record == 0,
+        return graphs
+    
 
-                # Misbehaving CSW server returning a next record > matched count
-                next_record > matched_count,
 
-                # No results returned already
-                returned_count == 0,
+class CswIsoXsltDcatBackend(DcatBackend):
+    display_name = 'CSW-ISO-XSLT-DCAT'
 
-                # Current next record is lower than previous one
-                next_record < start,
+    ISO_SCHEMA = 'http://www.isotc211.org/2005/gmd'
 
-                # Enough items have been harvested already
-                self.max_items and len(self.job.items) >= self.max_items
-            )
+    def get_format(self):
+        # TODO: should we redefine get_format here?
+        return 'xml'
+
+    def parse_graph(self, url: str, fmt: str) -> List[Graph]:
+        '''
+        Parse CSW graph querying ISO schema.
+        Use SEMIC GeoDCAT-AP XSLT to map it to a correct version.
+        See https://github.com/SEMICeu/iso-19139-to-dcat-ap for more information on the XSLT.
+        '''
+
+        # The XSLT transform failed on some catalog with "Cannot resolve URI https://ogc.geo-ide.developpement-durable.gouv.fr/csw/all-dataset?REQUEST=GetRecordById&SERVICE=CSW&VERSION=2.0.2&RESULTTYPE=results&elementSetName=full&TYPENAMES=gmd:MD_Metadata&OUTPUTSCHEMA=http://www.isotc211.org/2005/gmd&ID=fr-120066022-jdd-9bffd34b-916b-4dfd-a446-b7c5ec05291a", this EmptyResolver disable something (not sure what) and allow to continue fetching the catalog.
```

Comment:
Je n'arrive pas à reproduire le problème avec un [simple script](https://gist.github.com/streino/156d51c67501c40fc9794335b781a674) qui itère sur le catalogue "problématique" et applique le xslt.

J'ai pourtant bien l'item en question et l'URL qui provoque l'exception : 
```
$ grep 'fr-120066022-jdd-9bffd34b-916b-4dfd-a446-b7c5ec05291a' out-dcat-*
out-dcat-3.xml:        <dct:identifier rdf:datatype="http://www.w3.org/2001/XMLSchema#string">fr-120066022-jdd-9bffd34b-916b-4dfd-a446-b7c5ec05291a</dct:identifier>
out-dcat-3.xml:      <foaf:Document rdf:about="http://catalogue.geo-ide.developpement-durable.gouv.fr/catalogue/srv/fre/catalog.search#/metadata/fr-120066022-jdd-9bffd34b-916b-4dfd-a446-b7c5ec05291a">
out-dcat-3.xml:      <foaf:Document rdf:about="https://ogc.geo-ide.developpement-durable.gouv.fr/csw/all-dataset?REQUEST=GetRecordById&amp;SERVICE=CSW&amp;VERSION=2.0.2&amp;RESULTTYPE=results&amp;elementSetName=full&amp;TYPENAMES=gmd:MD_Metadata&amp;OUTPUTSCHEMA=http://www.isotc211.org/2005/gmd&amp;ID=fr-120066022-jdd-9bffd34b-916b-4dfd-a446-b7c5ec05291a">
```

@ThibaudDauce tu avais récupéré plus d'info que le "XSLTApplyError: Cannot resolve URI" ?

Reply:
Il semble que je produise bien le problème. J'ai repris ton script en [requêtant cette fois précisément une fiche](https://gist.github.com/maudetes/e14d24ecc04e82c64dacfb76e2d638d4) qui pose problème.
L'erreur a lieu sur `<srv:operatesOn uuidref="fr-120066022-jdd-baf93563-9b63-45b7-a458-16c011f946b3" xlink:href="https://ogc.geo-ide.developpement-durable.gouv.fr/csw/all-dataset?REQUEST=GetRecordById&amp;SERVICE=CSW&amp;VERSION=2.0.2&amp;RESULTTYPE=results&amp;elementSetName=full&amp;TYPENAMES=gmd:MD_Metadata&amp;OUTPUTSCHEMA=http://www.isotc211.org/2005/gmd&amp;ID=fr-120066022-jdd-baf93563-9b63-45b7-a458-16c011f946b3"/>`.
Je propose qu'on garde le fix apporté par @ThibaudDauce pour avancer.

Reply:
@maudetes La fiche [fr-120066022-srv-3e760aa6-99fc-44b4-9432-a32c3d31adb6](https://ogc.geo-ide.developpement-durable.gouv.fr/csw/csw-dreal-ge?REQUEST=GetRecordById&SERVICE=CSW&VERSION=2.0.2&RESULTTYPE=results&elementSetName=full&TYPENAMES=gmd:MD_Metadata&OUTPUTSCHEMA=http://www.isotc211.org/2005/gmd&ID=fr-120066022-srv-3e760aa6-99fc-44b4-9432-a32c3d31adb6) que tu requêtes est une fiche de service, pas de jeu de données.

Les points de moissonnages - que ce soit ceux de Géo-IDE ou ceux des autres catalogues - ne devraient en principe pas exposer de fiches de services, puisque data.gouv.fr ne les gère actuellement pas du tout. En ce qui concerne Géo-IDE, ils ont l'habitude de ce filtrage.

L'élément `srv:operatesOn` contient en principe une référence à la fiche de métadonnées du jeu de données mis à disposition par le service (et plus précisément à son élément `gmd:MD_DataIdentification`).
_Références : [GeoDCAT v2.0](https://semiceu.github.io/GeoDCAT-AP/releases/2.0.0/#coupled-resource---not-in-iso19115-core) ; [Guide technique INSPIRE](https://github.com/INSPIRE-MIF/technical-guidelines/blob/main/metadata/metadata-iso19139/metadata-iso19139.adoc#4124-linking-to-provided-data-sets-using-coupled-resource)._ 

Visiblement, le problème est que le XSLT [tente d'aller interroger](https://github.com/SEMICeu/iso-19139-to-dcat-ap/blob/34b64006053087e7b2674604312419d09a7af826/iso-19139-to-dcat-ap.xsl#L2107) cette fiche de métadonnées distante et échoue - ne serait que parce que, par défaut, [`lxml` bloque ces requêtes](https://lxml.de/resolvers.html#xml-catalogs), et ce n'est peut-être pas le seul verrou.

Il y a un paramètre qui permet d'empêcher l'interrogation des ressources liées lors de l'exécution du XSLT :  [`$CoupledResourceLookup`](https://github.com/SEMICeu/iso-19139-to-dcat-ap/blob/master/documentation/HowTo.md#parameter-coupledresourcelookup).

Et, effectivement, je n'ai plus l'erreur en remplaçant cette commande de ton script 
```python
dcat = transform(iso)
```

par 
```python
dcat = transform(iso, CoupledResourceLookUp="'disabled'")
```

Je recommanderais d'utiliser ce paramètre et de désactiver le correctif pour la suite des tests, dans la mesure où il pourrait empêcher de détecter d'autres problèmes.

Reply:
Peut-être aussi ajouter un filtre à la requête CSW pour ne récupérer que les datasets (ou à l'inverse exclure les services) ?

```xml
<csw:GetRecords ...>
  <csw:Query ...>
    ...
    <csw:Constraint version="1.1.0">
      <ogc:Filter xmlns:ogc="http://www.opengis.net/ogc">
        <ogc:PropertyIsEqualTo>
          <ogc:PropertyName>dc:type</ogc:PropertyName>
          <ogc:Literal>dataset</ogc:Literal>
        </ogc:PropertyIsEqualTo>
      </ogc:Filter>
    </csw:Constraint>
  </csw:Query>
</csw:GetRecords>
```

Reply:
Merci @alhyss pour l'enquête et le paramètre ! Comme ça a l'air fonctionnel sur le catalogue de DREAL Grand Est, on fait la modification !
@streino, on peut rajouter une requête pour filtrer les datasets (je pensais que c'était le cas avec la requête actuelle).
Par contre, je veux bien validation que la requête sera fonctionnelle sur toutes les plateformes ?

Reply:
Dans [cette version](https://portal.ogc.org/files/80534) des specs CSW 2.0.2, c'est la partie 8.3 en page 90 :
_Pointeur de @alhyss, merci !_

> 8.3 Query facilities
> ...
> OGC Filter Encoding is an XML based encoding of the OGC_Common Query Language. This query language (Version 1.1.0) must be supported by all catalogue instances of this profile in order to support search interoperability. In addition, the plain textual encoding of CQL version 2.0 may be supported. ...
> 
> The following conditions must be met by the supported query language:
> - support for the comparison operators:
>   - EqualTo
>   - ...
> - support for the following expressions:
>   - property name
>   - literal
> - ...

En pratique, on avait des filtres OGC sur quasi tous les [moissonnages de notre POC CKAN](https://github.com/ecolabdata/guichetdonnees-public/blob/main/moissonnages.json) et nous n'avons pas rencontré de problèmes.

Sinon pour être 100% certain il faudra passer par une première requête [GetCapabilities](https://catalogue.datara.gouv.fr/geonetwork/srv/fre/csw-dreal_opendata?service=CSW&version=2.0.2&namespace=xmlns:csw=http://www.opengis.net/cat/csw&request=GetCapabilities) avant le GetRecords : 

```
<csw:Capabilities ...>
  ...
  <ows:OperationsMetadata>
    ...
    <ows:Operation name="GetRecords">
      ...
      <ows:Parameter name="CONSTRAINTLANGUAGE">
        <ows:Value>FILTER</ows:Value>
        ...

  <ogc:Filter_Capabilities>
    ...
    <ogc:Scalar_Capabilities>
      ...
      <ogc:ComparisonOperators>
        <ogc:ComparisonOperator>EqualTo</ogc:ComparisonOperator>
        ...
```

Reply:
Merci :pray: J'ai rajouté le filtre sur la requête (sans utilisé `GetCapabilities` pour l'instant). Si vous les aviez utilisé de votre côté sans soucis, ça me paraît correct.